File: src/main/java/org/droidmate/analyzer/evaluation/EvaluationStrategyBuilder.java
package org.droidmate.analyzer.evaluation;

import org.apache.commons.lang3.NotImplementedException;
import org.droidmate.analyzer.exploration.IScenario;

/**
 * Builder pattern for EvaluationStrategy
 */
public class EvaluationStrategyBuilder {
    private EvaluationType type;
    private double threshold;

    public EvaluationStrategyBuilder(EvaluationType type, double threshold){
        this.type = type;
        this.threshold = threshold;

        assert this.type != null;
    }

    public IEvaluationStrategy build(IScenario initialExpl){
        if (initialExpl == null)
            return new InitialExplStrategy();

        switch (this.type){
            case DidNotCrash:
                return new DidNotCrash();
            case SimilarSize:
                return new SimilarExplorationSize(initialExpl, this.threshold);
            case SimilarApis:
                return new SimilarApis(initialExpl, this.threshold);
            case AlwaysValid:
                return new AlwaysValid(initialExpl, this.threshold);
            default:
                throw new NotImplementedException("Evaluation type required");
        }
    }
}

File: src/main/kotlin/org/droidmate/analyzer/evaluation/EvaluationStrategyBuilder.kt
package org.droidmate.analyzer.evaluation

import org.apache.commons.lang3.NotImplementedException
import org.droidmate.analyzer.exploration.IScenario

/**
 * Builder pattern for EvaluationStrategy
 */
class EvaluationStrategyBuilder(private val type: EvaluationType?, private val threshold: Double) {

    init {

        assert(this.type != null)
    }

    fun build(initialExpl: IScenario?): IEvaluationStrategy {
        if (initialExpl == null)
            return InitialExplStrategy()

        when (this.type) {
            EvaluationType.DidNotCrash -> return DidNotCrash()
            EvaluationType.SimilarSize -> return SimilarExplorationSize(initialExpl, this.threshold)
            EvaluationType.SimilarApis -> return SimilarApis(initialExpl, this.threshold)
            EvaluationType.AlwaysValid -> return AlwaysValid(initialExpl, this.threshold)
            else -> throw NotImplementedException("Evaluation type required")
        }
    }
}
